### DISCUSSION TITLE
Table specialization with inheritance

### DISCUSSION THREAD

**Discussion by claria** [sqlalchemy/sqlalchemy#10612] [Usage Questions]
Dear community,

our current use case is the following:
- We have a range of customers which are separated by a distinct database for each customer.
- We use separate engine/connections/sessions to connect to each database
- All customers have the same set of tables with exactly the same schema
- A very small number of customers has additional columns in some of their tables.

Currently we have organized the ORM like the following

```
Base = declarative_base()
```

The common set of columns/attribute is available and used by the specification of the CommonTable:

```
class CommonTable(Base):
    __tablename__ = "common_table"

    id= Column(Integer, primary_key=True)
   # ... about 20 more common columns
```

The table, which is extended for a signle customer with an additional column, inherits from the common table.

```
class ExtendedTableCustomerX(CommonTable):

    additional_column= Column(String(100)) # additional column which makes only sense for customer x
```

In the same running process, both variants are imported since for example a special function needs to access the additional column
if it is run in the context of "customer x".

However, this does not work correctly. Since both of them derive from the same Base and use the same metadata object, the additional column is added to the metadata of the "common_table" and this leads to problems when used, e.g. column is expected by a query built by sqlalchemy.

I have the feeling, that we have a fundamental error here in our usage design and wanted to ask what is the general recommendation for such a usecase. 
- I read the documentation about inheritance but do not think that the described usecases fit our case. 
- I am wondering if we need to have separate metadata objects which are not attached to the Base, but to the customer-database-engine? Is there a concept like this?

Thank you for your help to guide us in the correct direction!

Database: MSSQL Server (Azure)
SQLAlchemy: 1.4

**Comment by zzzeek** [comment]
thanks for the clear description of the problem.   I agree ORM inheritance is not appropriate here since you are dealing with a concrete class per customer, and additionally the desire to have repeatable table definitions can be suited using [mixins (1.4 doc version)](https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html).

Easier to demonstrate so below is a POC using two separate sqlite engines per "customer", using a common mixin for the columns that are common:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
Base = declarative_base()


class CommonTableMixin:
    __tablename__ = "common_table"

    id= Column(Integer, primary_key=True)

    x = Column(Integer)
    y = Column(Integer)
    z = Column(Integer)

class CommonTable(CommonTableMixin, Base):
    pass

class ExtendedTableCustomerX(CommonTableMixin, Base):
    metadata = MetaData()

    additional_column= Column(String(100)) # additional column which makes only sense for customer x

customer_one_engine = create_engine("sqlite://", echo=True, logging_name="customer_one")
customer_X_engine = create_engine("sqlite://", echo=True, logging_name="customer_X")

Base.metadata.create_all(customer_one_engine)
ExtendedTableCustomerX.metadata.create_all(customer_X_engine)


with Session(customer_one_engine) as sess:
    c1 = CommonTable(x=1, y=2, z=3)
    sess.add(c1)
    sess.commit()

    assert sess.scalar(select(CommonTable)) is c1

with Session(customer_X_engine) as sess:
    e1 = ExtendedTableCustomerX(x=1, y=2, z=3, additional_column="bonus content")
    sess.add(e1)
    sess.commit()

    assert sess.scalar(select(ExtendedTableCustomerX)) is e1
```

**Comment by claria** [comment]
Dear @zzzeek 

thank you very much for your detailed answer and the example code. I think this is the way to go as it is very clear. I am facing however one problem. 

The additional metadata in the extended table knows only this one table. If the CommonTableMixin has some relationship to another table CommonRelatedTable, this will give an error.

To test it, i overwrote the metadata.tables with the dict from Base.metadata.tables+metadata.tables. After this, it works, but this is of course .... Is there a way to specify metadata as "all tables from Base.metadata but CommonTable should be replaced by ExtendedTableCustomerX"?

```python
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import ForeignKey
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.orm import declared_attr
Base = declarative_base()

class CommonRelatedTable(Base):
    __tablename__ = "common_related_table"
    id= Column(Integer, primary_key=True)
    
    
class CommonTableMixin:
    __tablename__ = "common_table"

    id= Column(Integer, primary_key=True)

    @declared_attr
    def related_table_id(self):
        return Column(Integer, ForeignKey('common_related_table.id'))

    x = Column(Integer)
    y = Column(Integer)
    z = Column(Integer)

class CommonTable(CommonTableMixin, Base):
    pass

metadata = MetaData()

class ExtendedTableCustomerX(CommonTableMixin, Base):
    metadata = metadata

    additional_column= Column(String(100)) # additional column which makes only sense for customer x

# stupid way to extend metadata
tmp_tables = dict(**Base.metadata.tables)
tmp_tables.update(metadata.tables)
metadata.tables = tmp_tables

customer_one_engine = create_engine("sqlite://", echo=True, logging_name="customer_one")
customer_X_engine = create_engine("sqlite://", echo=True, logging_name="customer_X")

Base.metadata.create_all(customer_one_engine)
ExtendedTableCustomerX.metadata.create_all(customer_X_engine)

print('customer one')
with Session(customer_one_engine) as sess:
    r1 = CommonRelatedTable(id=0)
    sess.add(r1)
    c1 = CommonTable(x=1, y=2, z=3, related_table_id=0)
    sess.add(c1)
    sess.commit()

    sess.scalar(select(CommonRelatedTable))
    sess.scalar(select(CommonTable))

    assert sess.scalar(select(CommonTable)) is c1

print('customer x')
with Session(customer_X_engine) as sess:
    r1 = CommonRelatedTable(id=0)
    sess.add(r1)
    e1 = ExtendedTableCustomerX(x=1, y=2, z=3, related_table_id=0, additional_column="bonus content")
    sess.add(e1)
    sess.commit()
    
    
    sess.scalar(select(CommonRelatedTable))
    sess.scalar(select(CommonTable))
    sess.scalar(select(ExtendedTableCustomerX))
    
    assert sess.scalar(select(ExtendedTableCustomerX)) is e1
```

**Reply by zzzeek** [reply]
you dont want to have copies of all those tables around, easiest way would be to make the ForeignKeys on commontable mixin use the real column in terms of the base metadata

```py
    
class CommonTableMixin:
    __tablename__ = "common_table"

    id= Column(Integer, primary_key=True)

    @declared_attr
    def related_table_id(self):
        return Column(Integer, ForeignKey(Base.metadata.tables['common_related_table'].c.id))
```

just tested and this works
